### PR TITLE
WinUI 3 Preview 1

### PR DESCRIPTION
**Pull request by asklar**
- Update WinUI 3 flavor to target Preview 1
- the CI had a bug where we were not passing additionalInitArguments, instead we were passing the string `$[ coalesce($(additionalInitArguments), '') ]$`
- the CLI used to not reject unknown arguments/commands, which wrongly made the previous issue go unnoticed, so I fixed that
- Made the CLI use the WinUI 3 version from WinUI.props

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/5309)

### CODE REVIEW

**Comment by dannyvv** [comment]
Don't forget to also update `vnext\local-cli\generator-windows\index.js`  line 110: `version: options.useWinUI3 ? '3.0.0-alpha.200210.0' : '2.3.191129002',`

File: vnext/ReactUWP/Views/FlyoutViewManager.cpp [hunk 1]
```diff
@@ -226,10 +226,10 @@ void FlyoutShadowNode::createView() {
       });
 
   // Set XamlRoot on the Flyout to handle XamlIsland/AppWindow scenarios.
-  if (auto flyoutBase6 = m_flyout.try_as<winrt::IFlyoutBase6>()) {
+  if (Is19H1OrHigher()) {
```
Comment (dannyvv) [inline vnext/ReactUWP/Views/FlyoutViewManager.cpp:229]: >Is19H1OrHigher [](start = 6, length = 14)

Question: Wasn't the whole guidance from the windows sdk team when win10 released to stop performing version checks and switch to capability checks?

Comment (asklar) [reply vnext/ReactUWP/Views/FlyoutViewManager.cpp:229]: the implementation of `Is19H1OrHigher` is a capability check (checks whether a certain contract is present)

File: vnext/Microsoft.ReactNative/ReactApplication.h [hunk 2]
```diff
@@ -19,19 +25,32 @@ namespace winrt::Microsoft::ReactNative::implementation {
...
-          composable,
-          composing,
-          xaml::IApplicationOverrides,
-          xaml::IApplicationOverrides2,
-          I...>,
-      impl::require<D, xaml::IApplication, xaml::IApplication2, xaml::IApplication3>,
-      impl::base<D, xaml::Application>,
-      xaml::IApplicationOverridesT<D>,
-      xaml::IApplicationOverrides2T<D> {
+struct __declspec(empty_bases) NoDefaultCtorReactApplication_base : implements<
```
Comment (rectified95) [inline vnext/Microsoft.ReactNative/ReactApplication.h:28]: why not just have NoDefaultCtorReactApplication_base defined twice with the appropriate args, wrapped inside a #ifdef clause. imo much more readable

File: vnext/ReactUWP/Views/SIPEventHandler.cpp [hunk 3]
```diff
@@ -47,11 +47,15 @@ void SIPEventHandler::InitializeCoreInputView() {
...
 
+#ifndef USE_WINUI3
     if (Is19H1OrHigher()) {
       // 19H1 and higher supports island scenarios
       auto uiElement(xamlView.as<xaml::UIElement>());
       m_coreInputView = winrt::CoreInputView::GetForUIContext(uiElement.UIContext());
-    } else {
+    } else
+#endif
+    /// TODO: Figure out how to use SIP in WinUI 3 and island
```
Comment (rectified95) [inline vnext/ReactUWP/Views/SIPEventHandler.cpp:57]: create an issue for this?

File: vnext/ReactUWP/Views/TextInputViewManager.cpp [hunk 4]
```diff
@@ -16,10 +16,25 @@
 
 #include <IReactInstance.h>
 
+#ifdef USE_WINUI3
+namespace winrt {
+namespace Microsoft {
```
Comment (rectified95) [inline vnext/ReactUWP/Views/TextInputViewManager.cpp:21]: use nested namespaces?

File: change/react-native-windows-init-2020-06-24-03-38-49-winui3preview1.json [hunk 5]
```diff
@@ -0,0 +1,8 @@
+{
+  "type": "patch",
+  "comment": "make cli strict and pass true to trailing bool param",
```
Comment (rectified95) [inline change/react-native-windows-init-2020-06-24-03-38-49-winui3preview1.json:3]: What's our strategy for multiple change files? Do we want them, or should we just summarize changes in one, and rebase locally?

Comment (asklar) [reply change/react-native-windows-init-2020-06-24-03-38-49-winui3preview1.json:3]: @rectified95 these are changes in different packages so we need one change file for each

File: vnext/ReactUWP/Views/FlyoutViewManager.cpp (hunk 1 above)
Comment (kmelmon) [reply vnext/ReactUWP/Views/FlyoutViewManager.cpp:229]: When we're running in WinUI mode, these XAML APIs should be present even if we're running downlevel.  It seems like the helper functions could be more permissive.  Am I missing something?

---
In reply to: [444601167](https://github.com/microsoft/react-native-windows/pull/5309#discussion_r444601167) [](ancestors = 444601167)

File: vnext/ReactUWP/Views/SIPEventHandler.cpp (hunk 3 above)
Comment (asklar) [reply vnext/ReactUWP/Views/SIPEventHandler.cpp:57]: #5333

File: vnext/Microsoft.ReactNative/ReactApplication.h [hunk 6]
```diff
@@ -19,19 +25,32 @@ namespace winrt::Microsoft::ReactNative::implementation {
...
+                                                                        D,
+                                                                        Microsoft::ReactNative::ReactApplication,
+                                                                        composable,
+                                                                        composing,
+                                                                        xaml::IApplicationOverrides,
+#ifndef USE_WINUI3
+                                                                        xaml::IApplicationOverrides2,
```
Comment (kmelmon) [inline vnext/Microsoft.ReactNative/ReactApplication.h:35]: These sorts of differences are concerning... Do you have a sense of how widespread it is?

Comment (asklar) [reply vnext/Microsoft.ReactNative/ReactApplication.h:35]: WinUI 3 flattened the API so you'll see interfaces like `IUIElement7`, etc. that don't exist in WinUI 3 anymore, so it's pretty widespread.

File: vnext/ReactUWP/Views/FlyoutViewManager.cpp (hunk 1 above)
Comment (asklar) [reply vnext/ReactUWP/Views/FlyoutViewManager.cpp:229]: @kmelmon actually...yes, I think that's right. I just need to do the same thing I did in the other places of aliasing the IFooN to IFoo for winui3.

File: vnext/ReactUWP/Views/TextInputViewManager.cpp (hunk 4 above)
Comment (asklar) [reply vnext/ReactUWP/Views/TextInputViewManager.cpp:21]: @rectified95 thanks - apparently that's possible now as of C++17 ('twasn't before)

File: vnext/ReactUWP/Views/FlyoutViewManager.cpp (hunk 1 above)
Comment (kmelmon) [reply vnext/ReactUWP/Views/FlyoutViewManager.cpp]: Looks good thanks!

---
In reply to: [445265358](https://github.com/microsoft/react-native-windows/pull/5309#discussion_r445265358) [](ancestors = 445265358)

File: vnext/Microsoft.ReactNative/ReactApplication.h (hunk 6 above)
Comment (kmelmon) [reply vnext/Microsoft.ReactNative/ReactApplication.h]: Maybe we should consolidate where possible (eg have a header file with the interface aliases).  It won't be possible in all cases (like this one), just trying to find ways to reduce/share instances of #ifdef USE_WINUI3.

---
In reply to: [445211208](https://github.com/microsoft/react-native-windows/pull/5309#discussion_r445211208) [](ancestors = 445211208)

**Pull request merged**